ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -0,0 +1,188 @@
+initMempoolEnv chainDB = do
+    isVar <- atomically $ newTVar initInternalState
+    return $ MempoolEnv chainDB isVar
+
+{-------------------------------------------------------------------------------
+  Implementation
+-------------------------------------------------------------------------------}
+
+-- TODO: This may return some transactions as invalid that aren't new. Remove?
+implAddTxs :: forall m blk hdr. (MonadSTM m, StandardHash blk, ApplyTx blk)
+           => MempoolEnv m blk hdr
+           -> [GenTx blk]
+           -> m [(GenTx blk, LedgerError blk)]
+implAddTxs mpEnv@MempoolEnv{..} txs = atomically $ do
+    ValidationResult{..} <- validateNew <$> validateIS mpEnv
+    writeTVar mpEnvStateVar IS { isTxs = vrValid
+                               , isTip = vrBefore
+                               }
+    return vrInvalid
+  where
+    validateNew :: ValidationResult blk ->  ValidationResult blk
+    validateNew = extendsVR $ map (NotPrevApplied, ) txs
+
+implGetTxs :: (MonadSTM m, StandardHash blk, ApplyTx blk)
+           => MempoolEnv m blk hdr
+           -> STM m (Seq (GenTx blk ))
+implGetTxs mpEnv@MempoolEnv{..} = do
+    ValidationResult{..} <- validateIS mpEnv
+    -- TODO (maybe): log the transactions that we silently discard
+    writeTVar mpEnvStateVar IS { isTxs = vrValid
+                               , isTip = vrBefore
+                               }
+    return vrValid
+
+{-------------------------------------------------------------------------------
+  Validation
+-------------------------------------------------------------------------------}
+
+data ValidationResult blk = ValidationResult {
+    -- | The tip of the chain before applying these transactions
+    vrBefore  :: ChainHash blk
+
+    -- | The transactions that were found to be valid (oldest to newest)
+  , vrValid   :: Seq (GenTx blk)
+
+    -- | The state of the ledger after 'vrValid'
+    --
+    -- NOTE: This is intentionally not a strict field, so that we don't
+    -- evaluate the final ledger state if we don't have to.
+  , vrAfter   :: LedgerState blk
+
+    -- | The transactions that were invalid, along with their errors
+    --
+    -- Order not guaranteed
+  , vrInvalid :: [(GenTx blk, LedgerError blk)]
+  }
+
+-- | Initialize 'ValidationResult' from a ledger state and a list of
+-- transactions /known/ to be valid in that ledger state
+initVR :: forall blk. ApplyTx blk
+       => Seq (GenTx blk)
+       -> (ChainHash blk, LedgerState blk)
+       -> ValidationResult blk
+initVR = \knownValid (tip, st) -> ValidationResult {
+      vrBefore  = tip
+    , vrValid   = knownValid
+    , vrAfter   = afterKnownValid (Foldable.toList knownValid) st
+    , vrInvalid = []
+    }
+  where
+    afterKnownValid :: [GenTx blk] -> LedgerState blk -> LedgerState blk
+    afterKnownValid []       st = st
+    afterKnownValid (tx:txs) st =
+        case runExcept (applyTx PrevApplied tx st) of
```

Comment:
It looks like we have two different interpretations of `PrevApplied` with `applyTx` and it'd probably be better to fully distinguish them.

In this case we mean that it's previously been applied **in this ledger state** and so it should not be possible to fail.

In the other case we mean that it's previously been applied **in some state** and so it is totally possible to fail, but should not fail (and doesn't need to check) conditions that are independent of the ledger state (like tx sigs).

Distinguishing the cases allows the ledger rules to optimise the first case by skipping all validation checks and only doing the state changes.

Reply:
`PrevApplied` is only intended as a hint to the ledger layer ("this has been previously applied"). The fact that in this particular case we know it must even be valid _for this particular ledger state_ is an outlier; `PrevApplied` _still_ means the same ("you can skip the checks of signatures"), it just so happens that _we_ know that we are validating this against the _very same_ ledger state that we did before. If the mempool would remember the ledger state at the end of the transactions, this computation and `error` case would not be needed. The more interesting use of `PrevApplied` is in `validateIS`, where we are re-validating the existing transactions against a _different_ ledger state. 

So I'm not sure that there is much to distinguish here? What we cannot express, currently, is "compute the new transaction state, and by the way, we have computed this transaction against this very same state before and we know it won't fail". But I'm not sure that the ledger layer could _do_ very much with that information?